bridge-history-api/internal/logic/watcher_event_parser.go

```diff
@@ -0,0 +1,476 @@
+				L1BlockNumber: vlog.BlockNumber,
+				L1TxHash:      vlog.TxHash.String(),
+				TxStatus:      int(orm.TxStatusTypeRelayedFailed),
+			})
+		}
+	}
+	return l1DepositMessages, l1RelayedMessages, nil
+}
+
+// ParseL1BatchEventLogs parses L1 watched batch events.
+func ParseL1BatchEventLogs(ctx context.Context, logs []types.Log, blockTimestampsMap map[uint64]uint64, client *ethclient.Client) ([]*orm.BatchEvent, error) {
+	var l1BatchEvents []*orm.BatchEvent
+	for _, vlog := range logs {
+		switch vlog.Topics[0] {
+		case backendabi.L1CommitBatchEventSig:
+			event := backendabi.L1CommitBatchEvent{}
+			if err := utils.UnpackLog(backendabi.IScrollChainABI, &event, "CommitBatch", vlog); err != nil {
+				log.Warn("Failed to unpack CommitBatch event", "err", err)
+				return nil, err
+			}
+			commitTx, isPending, err := client.TransactionByHash(ctx, vlog.TxHash)
+			if err != nil || isPending {
+				log.Warn("Failed to get commit Batch tx receipt or the tx is still pending", "err", err)
+				return nil, err
+			}
+			startBlock, endBlock, err := utils.GetBatchRangeFromCalldata(commitTx.Data())
+			if err != nil {
+				log.Warn("Failed to get batch range from calldata", "hash", commitTx.Hash().String(), "height", vlog.BlockNumber)
+				return nil, err
+			}
+			l1BatchEvents = append(l1BatchEvents, &orm.BatchEvent{
+				BatchStatus:      int(orm.BatchStatusTypeCommitted),
+				BatchIndex:       event.BatchIndex.Uint64(),
+				BatchHash:        event.BatchHash.String(),
+				StartBlockNumber: startBlock,
+				EndBlockNumber:   endBlock,
+			})
+		case backendabi.L1RevertBatchEventSig:
+			event := backendabi.L1RevertBatchEvent{}
+			if err := utils.UnpackLog(backendabi.IScrollChainABI, &event, "RevertBatch", vlog); err != nil {
+				log.Warn("Failed to unpack RevertBatch event", "err", err)
+				return nil, err
+			}
+			l1BatchEvents = append(l1BatchEvents, &orm.BatchEvent{
+				BatchStatus: int(orm.BatchStatusTypeReverted),
+				BatchIndex:  event.BatchIndex.Uint64(),
+				BatchHash:   event.BatchHash.String(),
+			})
+		case backendabi.L1FinalizeBatchEventSig:
+			event := backendabi.L1FinalizeBatchEvent{}
+			if err := utils.UnpackLog(backendabi.IScrollChainABI, &event, "FinalizeBatch", vlog); err != nil {
+				log.Warn("Failed to unpack FinalizeBatch event", "err", err)
+				return nil, err
+			}
+			l1BatchEvents = append(l1BatchEvents, &orm.BatchEvent{
+				BatchStatus: int(orm.BatchStatusTypeFinalized),
+				BatchIndex:  event.BatchIndex.Uint64(),
+				BatchHash:   event.BatchHash.String(),
+			})
+		}
+	}
+	return l1BatchEvents, nil
+}
+
+// ParseL1MessageQueueEventLogs parses L1 watched message queue events.
+func ParseL1MessageQueueEventLogs(ctx context.Context, logs []types.Log, blockTimestampsMap map[uint64]uint64, client *ethclient.Client) ([]*orm.MessageQueueEvent, error) {
+	var l1MessageQueueEvents []*orm.MessageQueueEvent
+	for _, vlog := range logs {
+		switch vlog.Topics[0] {
+		case backendabi.L1QueueTransactionEventSig:
```

Comment:
duplicated

Reply:
Nice catch. it's a type, that should be `L1DropTransactionEvent`.